tokio/src/time/delay_queue.rs

```diff
@@ -724,6 +724,9 @@ impl<T> DelayQueue<T> {
             self.delay = None;
 
             if let Some(idx) = self.wheel.poll(&mut self.poll, &mut self.slab) {
+                if let Some(deadline) = self.next_deadline() {
+                    self.delay = Some(delay_until(deadline));
+                }
```

Comment:
This function is already a little tough to follow without understanding most of the `DelayQueue` implementation. 

Instead of putting another `next_deadline()` in here, how about we set it in line 724--something like this:

```
self.delay = self.next_deadline().map(|deadline| delay_until(deadline));

if let Some(idx) = self.wheel.poll(&mut self.poll, &mut self.slab) {
    return Poll::Ready(Some(Ok(idx)))
}

if self.delay.is_none() {
    return Poll::Ready(None)
}
```

This ensures the next deadline is already set before checking if another item can be returned or completing if the queue is done.

Reply:
I did think about that, however, as far as I could tell `next_deadline` returns the next instant in the wheel. If that's the case, we need to ensure that we `poll` the wheel first, since otherwise the value instant by `next_deadline` would be the same as the value we just polled, causing an unecesasry wakeup in the future.

I agree that `poll_idx` is a little convoluted, I'll see what I can do.

Reply:
Yep that makes sense. Cleanup looks good!